ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,413 @@ let order_f (_, vm_rec) =
+        ranked_hosts
+      |> fst
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, host not in it" ;
+      raise Not_found
+  in
+  VMGroupMap.update group
+    (fun _ ->
+      Some
+        (ranked_hosts
+        |> IntMap.update vm_cnt_on_host (Option.map (List.filter (( <> ) host)))
+        |> IntMap.update (vm_cnt_on_host + 1) (fun hs ->
+               Some (host :: Option.value ~default:[] hs)
+           )
+        )
+    )
+    grp_ranked_hosts
+
+(*****************************************************************************************************)
+(* Planning code follows                                                                             *)
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+    vms_increasing vm_can_boot_on_host grp_ranked_hosts mapping =
+  debug "vm_anti_affinity_spread_evenly_plan" ;
+  log_vms_increasing ~__context vms_increasing ;
+  log_hosts_increasing ~__context hosts_increasing ;
+  log_grp_ranked_hosts ~__context grp_ranked_hosts ;
+  match vms_increasing with
+  | [] ->
+      mapping
+  | (vm, vm_size) :: remaining_vms -> (
+      let host_selector hosts =
+        hosts_increasing
+        |> List.filter (fun (h, size) -> List.mem h hosts && size >= vm_size)
+        |> List.map fst
+        |> fst_host_can_run vm_can_boot_on_host vm
+      in
+      match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+      | Some (`AntiAffinity group) -> (
+          debug
+            "Spread evenly plan: try to plan for the anti-affinity VM (%s %s)."
+            (Ref.string_of vm)
+            (Db.VM.get_name_label ~__context ~self:vm) ;
+          try
+            let h =
+              Xapi_vm_helpers.select_host_from_ranked_groups vm host_selector
+                (VMGroupMap.find group grp_ranked_hosts
+                |> IntMap.bindings
+                |> List.map snd
+                )
+            in
+            let h_size =
+              Int64.sub
+                (List.assoc h hosts_increasing)
+                (List.assoc vm vms_increasing)
+            in
+            debug
+              "Spread evenly plan: choose the host with the least free memory \
+               which can run the vm: (%s %s)."
+              (Ref.string_of h)
+              (Db.Host.get_name_label ~__context ~self:h) ;
+
+            let hosts_increasing =
```

Comment:
Would a `HostMap` be better for this purpose? E.g.
```
HostMap.update h Option.map (fun free_mem -> free_mem - vm_size) hosts
```
And the sorting can be done in `host_selector`:
```
let free_mem h = Option.value ~default:0 (HostMap.find_opt h hosts) in
List.sort (fun h1 h2 -> compare (free_mem h1) (free_mem h2)) hs
```
(the name `hosts_increasing` and others like `vms_increasing` are too long)

Reply:
> And the sorting can be done in host_selector:
Then `host_selector` will have time complexity of `N * Log N` at least, and `host_selector` is called for each anti-affinity VMs to be evacuated during host evacuation.

In the current implementation, the hosts (of each rank) are kept sorted increasingly in terms of the host's free memory (sorted only once when initialize by `anti_affinity_grp_ranked_hosts`, and updated every time a VM is planned to a host by `update_anti_affinity_grp_ranked_hosts` with time complexity of `N`), and `host_selector` can finish with time complexity about O(1) (iterate to find the first host which can boot the VM).